File: baseunits/modules/FoOlSlide.pas
unit FoOlSlide;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread, Cloudflare, RegExpr, synautil;

implementation

const
  dirurl = '/directory/';
  dirurlreader = '/reader/directory/';
  dirurlfoolslide = '/foolslide/directory/';
  dirurlslide = '/slide/directory/';
  dirurlslideU = '/Slide/directory/';
  dirurlonline = '/online/directory/';
  dirurlhelvetica = '/r/directory/';    

function GETWithCookie(const AHTTP: THTTPSendThread; const AURL: String;
  const Module: TModuleContainer): Boolean;
begin
  if ((Module.Website = 'SeinagiAdultoFansub') or
      (Module.Website = 'TripleSevenScan'))
    and (Pos(dirurl, AURL) = 0)then
    Result := AHTTP.POST(AURL, 'adult=true')
  else
    Result := AHTTP.GET(AURL);
end;

function GetDirURL(const AWebsite: String): String;
begin
  if (AWebsite = 'GoManga') or
     (AWebsite = 'Jaiminisbox') or
     (AWebsite = 'TripleSevenScan') then
    Result := dirurlreader
  else
  if AWebsite = 'OneTimeScans' then
    Result := dirurlfoolslide
  else
  if (AWebsite = 'DejameProbar') or
     (AWebsite = 'MenudoFansub') or
     (AWebsite = 'NeoProjectScan') or
     (AWebsite = 'SolitarioNoFansub') then
    Result := dirurlslide
  else
  if AWebsite = 'SantosScan' then
    Result := dirurlslideU
  else
  if (AWebsite = 'Pzykosis666HFansub') or
     (AWebsite = 'SeinagiFansub') then
    Result := dirurlonline
  else
  if(AWebsite = 'HelveticaScans') then
    Result := dirurlhelvetica     
  else
    Result := dirurl;
end;

function GetDirectoryPageNumber(const MangaInfo: TMangaInformation;
  var Page: Integer; const WorkPtr: Integer; const Module: TModuleContainer): Integer;
var
  v: IXQValue;
  p: Integer;
begin
  Result := NET_PROBLEM;
  Page := 1;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  if GETWithCookie(MangaInfo.FHTTP, Module.RootURL + GetDirURL(Module.Website), Module) then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        with TRegExpr.Create do
          try
            Expression := '/(\d+)/$';
            for v in XPath('//*[@class="next"]/a/@href') do
            begin
              Exec(v.toString);
              if SubExprMatchCount > 0 then
              begin
                p := StrToIntDef(Match[1], -1);
                if p > Page then
                  Page := p;
              end;
            end;
          finally
            Free;
          end;
      finally
        Free;
      end;
  end;
end;

function GetNameAndLink(const MangaInfo: TMangaInformation;
  const ANames, ALinks: TStringList; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  v: IXQValue;
  s: String;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  s := Module.RootURL + GetDirURL(Module.Website);
  if AURL <> '0' then s += IncStr(AURL) + '/';
  if GETWithCookie(MangaInfo.FHTTP, s, Module) then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        for v in XPath('//div[@class="list series"]/div/div[@class="title"]/a') do begin
          ALinks.Add(v.toNode.getAttribute('href'));
          ANames.Add(v.toString);
        end;
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  v: IXQValue;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.FHTTP, MangaInfo.mangaInfo do
  begin
    url := FillHost(Module.RootURL, AURL);
    if GETWithCookie(MangaInfo.FHTTP, url, Module) then
    begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do
        try
          coverLink := XPathString('//div[@class="thumbnail"]/img/@src');
          if title = '' then title := XPathString('//h1[@class="title"]');
          authors := TrimLeftChar(XPathString(
            '//div[@class="info"]/*[contains(text(),"Author")]/following-sibling::text()[1]'), [':', ' ']);
          artists := TrimLeftChar(XPathString(
            '//div[@class="info"]/*[contains(text(),"Artist")]/following-sibling::text()[1]'), [':', ' ']);
          summary := TrimLeftChar(XPathString(
            '//div[@class="info"]/*[contains(text(),"Synopsis")]/following-sibling::text()[1]'), [':', ' ']);
          for v in XPath('//div[@class="list"]//div[@class="title"]/a') do
          begin
            chapterLinks.Add(v.toNode.getAttribute('href'));
            if v.toNode.getAttribute('title') <> '' then
              chapterName.Add(v.toNode.getAttribute('title'))
            else chapterName.Add(v.toString);
          end;
          InvertStrings([chapterLinks, chapterName]);
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  v: IXQValue;
  s: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.FHTTP, DownloadThread.Task.Container do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    if GETWithCookie(DownloadThread.FHTTP, FillHost(Module.RootURL, AURL), Module) then
    begin
      Result := True;
      with TXQueryEngineHTML.Create(Document) do
        try
          PageNumber := XPath('//div[@class="topbar_right"]//ul[@class="dropdown"]/li').Count;
          s := XPathString('//script[contains(.,"var pages")]');
          if s <> '' then begin
            s := GetBetween('var pages = ', ';', s);
            try
              ParseHTML(s);
              for v in XPath('json(*)()("url")') do
                PageLinks.Add(v.toString);
            except
            end;
          end;
        finally
          Free;
        end;
    end;
  end;
end;

function GetImageURL(const DownloadThread: TDownloadThread;
  const AURL: String; const Module: TModuleContainer): Boolean;
var
  s: String;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.Task.Container, DownloadThread.FHTTP do
  begin
    s := AURL;
    if DownloadThread.WorkId > 0 then
      s := AppendURLDelim(s) + 'page/' + IncStr(DownloadThread.WorkId);
    if GETWithCookie(DownloadThread.FHTTP, FillHost(Module.RootURL, s), Module) then
    begin
      Result := True;
      with TXQueryEngineHTML.Create(Document) do
        try
          PageLinks[DownloadThread.WorkId] := XPathString('//div[@id="page"]//img/@src');
        finally
          Free;
        end;
    end;
  end;
end;

procedure RegisterModule;

  function AddWebsiteModule(AWebsite, ARootURL: String): TModuleContainer;
  begin
    Result := AddModule;
    with Result do
    begin
      Website := AWebsite;
      RootURL := ARootURL;
      OnGetDirectoryPageNumber := @GetDirectoryPageNumber;
      OnGetNameAndLink := @GetNameAndLink;
      OnGetInfo := @GetInfo;
      OnGetPageNumber := @GetPageNumber;
      OnGetImageURL := @GetImageURL;
    end;
  end;

begin
  AddWebsiteModule('PowerManga', 'http://read.powermanga.org');

  AddWebsiteModule('Shoujosense', 'http://reader.shoujosense.com');
  AddWebsiteModule('GoManga', 'http://gomanga.co');
  AddWebsiteModule('OneTimeScans', 'http://otscans.com');
  AddWebsiteModule('SenseScans', 'http://reader.sensescans.com');
  AddWebsiteModule('Jaiminisbox', 'https://jaiminisbox.com');
  AddWebsiteModule('KireiCake', 'https://reader.kireicake.com');
  AddWebsiteModule('HelveticaScans', 'http://helveticascans.com');    

  //es-san
  AddWebsiteModule('DangoOnlineNoFansub', 'http://lector.dangolinenofansub.com');
  AddWebsiteModule('DejameProbar', 'http://dejameprobar.es');
  AddWebsiteModule('HoshinoFansub', 'http://manga.animefrontline.com');
  AddWebsiteModule('MangaWorksFansub', 'http://lector.mangaworksfansub.net');
  AddWebsiteModule('MasterPieceScans', 'http://reader.manga2me.net');
  AddWebsiteModule('MenudoFansub', 'http://www.menudo-fansub.com');
  AddWebsiteModule('NeoProjectScan', 'http://npscan.mangaea.net');
  AddWebsiteModule('Pzykosis666HFansub', 'http://pzykosis666hfansub.com');
  AddWebsiteModule('R15TeamScanlation', 'http://www.r15team.com');
  AddWebsiteModule('SantosScan', 'http://santosfansub.com');
  AddWebsiteModule('SeinagiFansub', 'http://seinagi.org');
  AddWebsiteModule('SeinagiAdultoFansub', 'http://adulto.seinagi.org');
  AddWebsiteModule('SolitarioNoFansub', 'http://snf.mangaea.net');
  AddWebsiteModule('TripleSevenScan', 'http://triplesevenscans.com')
end;

initialization
  RegisterModule;

end.

File: lua/modules/FoOlSlide.lua
local dirurl = '/directory/'
local dirurlreader = '/reader/directory/'
local dirurlfoolslide = '/foolslide/directory/'
local dirurlslide = '/slide/directory/'
local dirurlslideU = '/Slide/directory/'
local dirurlonline = '/online/directory/'
local dirurlhelvetica = '/r/directory/'
local dirurllector = '/lector/directory/'
local dirurlfsdir = '/fs/directory/'
local dirurlreaderlist = '/fs/reader/list/'

function getWithCookie(lurl)
  if http.get(lurl) then
    local x = TXQuery.Create(http.document)
    local s = x.xpathstring('//form//input[(@type="hidden") and (@name="adult")]/@value')
    if s:lower() == 'true' then
      http.reset()
      return http.post(lurl, 'adult=true')
    end
    return true
  else
    return false
  end
end

function getdirurl(website)
  local dirs = {
    ['Jaiminisbox'] = dirurlreader,
    ['DokiFansubs'] = dirurlreader,
    ['OneTimeScans'] = dirurl,
    ['DejameProbar'] = dirurlslide,
    ['MenudoFansub'] = dirurlslide,
    ['SolitarioNoFansub'] = dirurlslide,
    ['Pzykosis666HFansub'] = dirurlonline,
    ['SeinagiFansub'] = dirurlonline,
    ['HelveticaScans'] = dirurlhelvetica,
    ['RavensScans'] = dirurllector,
    ['NoraNoFansub'] = dirurllector,
    ['AntisenseScans'] = dirurlonline,
	['SenseScans'] = dirurlreader,
    ['MangaichiScan'] = dirurlfsdir,
    ['Riceballicious'] = dirurlreaderlist,
    ['Yuri-ism'] = dirurlslide,
    ['MangajinNoFansub'] = dirurllector,
    ['BunnysScans'] = '/read/directory/',
    ['CanisMajorScans'] = dirurlreader,
    ['HoshikuzuuScans'] = dirurl,
    ['YaoiIsLife'] = dirurlreader,
    ['FujoshiBitches'] = dirurlreader,
    ['TapTrans'] = dirurlfsdir,
    ['LoliVault'] = dirurlonline,
    ['Shoujohearts'] = dirurlreader
  }  
  if dirs[website] ~= nil then
    return dirs[website]
  else
    return dirurl
  end
end

function getinfo()
  local lurl = MaybeFillHost(module.rooturl, url)
  local result = net_problem
  if getWithCookie(lurl) then
    x = TXQuery.Create(http.document)
    mangainfo.coverlink = x.xpathstring('//div[@class="thumbnail" or contains(@class, "thumb")]/img/@src')
    if mangainfo.title == '' then
      mangainfo.title = x.xpathstring('//h1[@class="title"]')
    end
    if Pos('emailprotected', mangainfo.title) > 0 then
      mangainfo.title = Trim(SeparateLeft(x.xpathstring('//title'), '::'))
    end
    local cls = 'info'
    mangainfo.authors = string.gsub(
      x.xpathstring('//div[@class="'..cls..'"]/*[contains(text(),"Author")]/following-sibling::text()[1]'),
      '^[%s:]*', '')
    mangainfo.artists = string.gsub(
      x.xpathstring('//div[@class="'..cls..'"]/*[contains(text(),"Artist")]/following-sibling::text()[1]'),
      '^[%s:]*', '')
    mangainfo.summary = string.gsub(
      x.xpathstring('//div[@class="'..cls..'"]/*[contains(text(),"Synopsis")]/following-sibling::text()[1]'),
      '^[%s:]*', '')
    v = x.xpath('//div[@class="list"]//div[@class="title"]/a')
    for i = 1, v.count do
      v1 = v.get(i)
      mangainfo.chapterlinks.add(v1.getattribute('href'))
      if v1.getattribute('title') ~= '' then
        mangainfo.chapternames.add(v1.getattribute('title'))
      else
        mangainfo.chapternames.add(v1.ToString)
      end
    end
    InvertStrings(mangainfo.chapterlinks, mangainfo.chapternames)
    result = no_error
  end
  return result
end

function getinfo_ths()
  mangainfo.url=MaybeFillHost(module.RootURL, url)
  if getWithCookie(mangainfo.url) then
    local x=TXQuery.Create(http.document)
    mangainfo.title=x.xpathstring('//div[@id="series_right"]/h1')
    mangainfo.coverlink=MaybeFillHost(module.RootURL, x.xpathstring('//img[@class="series_img"]/@src'))
    mangainfo.authors = x.xpathstring('//ul[@class="series_left_data"]/li[contains(span, "Author")]/span[@class="value"]')
    mangainfo.artists = x.xpathstring('//ul[@class="series_left_data"]/li[contains(span, "Artist")]/span[@class="value"]')
    mangainfo.genres=x.xpathstringall('//ul[@class="series_left_data"]/li[contains(span, "Genre")]/span[@class="value"]/text()')
    mangainfo.status=MangaInfoStatusIfPos(x.xpathstring('//ul[@class="series_left_data"]/li[contains(span, "Status")]/span[@class="value"]'))
    mangainfo.summary = x.xpathstring('//div[@id="series_des"]')
    x.xpathhreftitleall('//div[@id="staff"]/div/a', mangainfo.chapterlinks, mangainfo.chapternames)
    InvertStrings(mangainfo.chapterlinks,mangainfo.chapternames)
    return no_error
  else
    return net_problem
  end
end

function taskstart()
  task.pagelinks.clear()
  task.pagenumber = 0
  return true
end

function getpagenumber()
  local result = false
  if getWithCookie(MaybeFillHost(module.rooturl, url)) then
    x = TXQuery.create(http.document)
    task.pagenumber = x.xpath('//div[@class="topbar_right"]//ul[@class="dropdown"]/li').count
    s = x.xpathstring('//script[contains(.,"var pages")]')
    if s ~= '' then
      s = GetBetween('var pages = ', ';', s)
      if Pos('atob("', s) > 0 then
         s = GetBetween('atob("', '")', s)
         s = DecodeBase64(s)  
      end
      x.parsehtml(s)
      v = x.xpath('json(*)()("url")')
      for i = 1, v.count do
        task.pagelinks.add(v.get(i).ToString)
      end
    end
    result = true
  end
  return result
end

function getpagenumber_jb()
  local result = false
  if getWithCookie(MaybeFillHost(module.rooturl, url)) then
    x = TXQuery.create(http.document)
    task.pagenumber = x.xpath('//div[@class="topbar_right"]//ul[@class="dropdown"]/li').count
    s = x.xpathstring('//script[contains(.,"[\'fromCharCode\',\'")]')
    if s ~= '' then
      s = GetBetween('[\'fromCharCode\',\'', '\'];', s)
      s = ExecJS("function decrypt(encrypted) {return encrypted.replace(/[a-zA-Z]/g,function (a) {return String.fromCharCode((a <= 'Z' ? 90 : 122) >= (a = a.charCodeAt(0) + 13) ? a : a - 26);});}; decrypt('" .. s .. "');")
      s = DecodeBase64(s)
      x.parsehtml(s)
      v = x.xpath('json(*)()("url")')
      for i = 1, v.count do
        task.pagelinks.add(v.get(i).ToString)
      end
    end
    result = true
  end
  return result
end

function getimageurl()
  local result = false
  local s = url
  if workid > 0 then
    s = AppendURLDelim(s) .. 'page/' .. (workid + 1)
  end
  if getWithCookie(MaybeFillHost(module.rooturl, s)) then
    x = TXQuery.create(http.document)
    task.pagelinks.set(workid, x.XPathString('//div[@id="page"]//img/@src'))
  end
  return result
end

function getdirectorypagenumber()
  local result = net_problem
  page = 1
  if getWithCookie(module.rooturl .. getdirurl(module.website)) then
    result = no_error
    x = TXQuery.create(http.document)
    v = x.xpath('//*[@class="next"]/a/@href')
    for i = 1, v.count do
      local s = tonumber(string.match(v.get(i).tostring, '/(%d+)/$'))
      if (s ~= nil) and (s > page) then
        page = s
      end
    end
  end
  return result
end

function getnameandlink()
  local result = net_problem
  local s = module.rooturl .. getdirurl(module.website)
  if url ~= '0' then
    s = s .. (tonumber(url) + 1) .. '/'
  end
  if getWithCookie(s) then
    result = no_error
    local x = TXQuery.create(http.document)
    if module.website == 'TwistedHelScans' then
      local v = x.xpath('//div[contains(@class, "series_card")]/a')
      for i = 1, v.count do
        local v1 = v.get(i)
        links.add(v1.getattribute('href'))
        names.add(x.xpathstring('span', v1))
      end
    else
      x.XpathHREFAll('//div[@class="list series"]/div/div[@class="title"]/a', links, names)
    end
  end
  return result
end

function AddWebsiteModule(name, url, category)
  local m = NewModule()
  m.website = name
  m.rooturl = url
  m.category = category
  m.ongetinfo = 'getinfo'
  m.OnTaskStart = 'taskstart'
  m.OnGetPageNumber = 'getpagenumber'
  m.OnGetImageURL = 'getimageurl'
  m.OnGetDirectoryPageNumber = 'getdirectorypagenumber'
  m.OnGetNameAndLink = 'getnameandlink'
  if name == 'TwistedHelScans' then m.ongetinfo = 'getinfo_ths'; end
  if name == 'Jaiminisbox' then m.OnGetPageNumber = 'getpagenumber_jb'; end
  return m
end

function Init()
  local cat = 'English-Scanlation'
  AddWebsiteModule('AntisenseScans', 'http://antisensescans.com', cat)
  AddWebsiteModule('BunnysScans', 'http://bns.shounen-ai.net', cat)
  AddWebsiteModule('CanisMajorScans', 'http://cm-scans.shounen-ai.net', cat)
  AddWebsiteModule('DeathTollScans', 'https://reader.deathtollscans.net', cat)
  AddWebsiteModule('DokiFansubs', 'https://kobato.hologfx.com', cat)
  AddWebsiteModule('EvilFlowers', 'http://reader.evilflowers.com', cat)
  AddWebsiteModule('ForgottenScans', 'http://reader.fos-scans.com', cat)
  AddWebsiteModule('FujoshiBitches', 'http://fujoshibitches.shounen-ai.net', cat)
  AddWebsiteModule('HelveticaScans', 'http://helveticascans.com', cat)
  AddWebsiteModule('HoshikuzuuScans', 'http://hoshiscans.shounen-ai.net', cat)
  AddWebsiteModule('IlluminatiManga', 'http://reader.manga-download.org', cat)
  AddWebsiteModule('Jaiminisbox', 'https://jaiminisbox.com', cat)
  AddWebsiteModule('KireiCake', 'https://reader.kireicake.com', cat)
  AddWebsiteModule('MangaichiScan', 'http://mangaichiscans.mokkori.fr', cat)
  AddWebsiteModule('OneTimeScans', 'https://reader.otscans.com', cat)
  AddWebsiteModule('PhoenixSerenade', 'https://reader.serenade.moe', cat)
  AddWebsiteModule('PowerManga', 'http://read.powermanga.org', cat)
  AddWebsiteModule('Riceballicious', 'http://riceballicious.info', cat)
  AddWebsiteModule('RoseliaScanlations', 'http://reader.roseliascans.com', cat)
  AddWebsiteModule('S2Scans', 'https://reader.s2smanga.com', cat)
  AddWebsiteModule('SeaOtterScans', 'https://reader.seaotterscans.com', cat)
  AddWebsiteModule('SenseScans', 'http://sensescans.com', cat)
  AddWebsiteModule('Shoujohearts', 'http://shoujohearts.com', cat)
  AddWebsiteModule('Shoujosense', 'http://reader.shoujosense.com', cat)
  AddWebsiteModule('SilentSkyScans', 'http://reader.silentsky-scans.net', cat)
  AddWebsiteModule('TapTrans', 'https://taptaptaptaptap.net', cat)
  AddWebsiteModule('TheCatScans', 'https://reader2.thecatscans.com', cat)
  AddWebsiteModule('TwistedHelScans', 'http://www.twistedhelscans.com', cat)
  AddWebsiteModule('VortexScans', 'https://reader.vortex-scans.com', cat)
  AddWebsiteModule('WorldThree', 'http://www.slide.world-three.org', cat)
  AddWebsiteModule('YaoiIsLife', 'http://yaoislife.shounen-ai.net', cat)
  AddWebsiteModule('Yuri-ism', 'https://www.yuri-ism.net', cat)
  
  -- es-sc
  cat = 'Spanish-Scanlation'
  AddWebsiteModule('KirishimaFansub', 'https://kirishimafansub.net', cat)
  AddWebsiteModule('LoliVault', 'https://lolivault.net', cat)
  AddWebsiteModule('Mangasubes', 'http://mangasubes.patyscans.com', cat)
  AddWebsiteModule('MenudoFansub', 'http://www.menudo-fansub.com', cat)
  AddWebsiteModule('NeoProjectScan', 'http://npscan.mangaea.net', cat)
  AddWebsiteModule('Nightow', 'http://nightow.net', cat)
  AddWebsiteModule('NoraNoFansub', 'https://www.noranofansub.com', cat)
  AddWebsiteModule('PCNet', 'http://pcnet.patyscans.com', cat)
  AddWebsiteModule('PatyScans', 'http://lector.patyscans.com', cat)
  AddWebsiteModule('Pzykosis666HFansub', 'https://pzykosis666hfansub.com', cat)
  AddWebsiteModule('RavensScans', 'http://ravens-scans.com', cat)
  AddWebsiteModule('SeinagiAdultoFansub', 'https://adulto.seinagi.org.es', cat)
  AddWebsiteModule('SeinagiFansub', 'https://seinagi.org.es', cat)
  AddWebsiteModule('SolitarioNoFansub', 'http://snf.mangaea.net', cat)
  AddWebsiteModule('TrueColorsScan', 'https://truecolorsscans.miocio.org', cat)
  AddWebsiteModule('XAnimeSeduccion', 'http://xanime-seduccion.com', cat)
end
